query-engine/core/src/executor/interactive_tx.rs

```diff
@@ -53,6 +89,423 @@ impl Display for TxId {
+                .await?;
+            results.push(Ok(result));
+        }
+
+        Ok(results)
+    }
+
+    pub async fn commit(&mut self) -> crate::Result<()> {
+        if let CachedTx::Open(_) = self.cached_tx {
+            let open_tx = self.cached_tx.as_open()?;
+            debug!("[{}] committing.", self.id.to_string());
+            open_tx.tx.commit().await?;
+            self.cached_tx = CachedTx::Committed;
+        }
+
+        Ok(())
+    }
+
+    pub async fn rollback(&mut self, was_timeout: bool) -> crate::Result<()> {
+        debug!("[{}] {was_timeout} rolling back", self.name());
+        if let CachedTx::Open(_) = self.cached_tx {
+            let open_tx = self.cached_tx.as_open()?;
+            open_tx.tx.rollback().await?;
+            if was_timeout {
+                debug!("[{}] Expired Rolling back", self.id.to_string());
+                self.cached_tx = CachedTx::Expired;
+            } else {
+                self.cached_tx = CachedTx::RolledBack;
+                debug!("[{}] Rolling back", self.id.to_string());
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn name(&self) -> String {
+        format!("itx-{:?}", self.id.to_string())
+    }
+}
+
+pub struct ITXClient {
+    send: Sender<TxOpRequest>,
+    tx_id: TxId,
+}
+
+impl ITXClient {
+    async fn commit(&self) -> crate::Result<()> {
+        let msg = self.send_and_receive(TxOpRequestMsg::Commit).await?;
+
+        if let TxOpResponse::Committed(resp) = msg {
+            debug!("[{}] COMMITTED {:?}", self.tx_id, resp);
+            resp
+        } else {
+            Err(self.handle_error(msg).into())
+        }
+    }
+
+    async fn rollback(&self) -> crate::Result<()> {
+        let msg = self.send_and_receive(TxOpRequestMsg::Rollback).await?;
+
+        if let TxOpResponse::RolledBack(resp) = msg {
+            resp
+        } else {
+            Err(self.handle_error(msg).into())
+        }
+    }
+
+    async fn execute(&self, operation: Operation, trace_id: Option<String>) -> crate::Result<ResponseData> {
+        let msg_req = TxOpRequestMsg::Single(operation, trace_id);
+        let msg = self.send_and_receive(msg_req).await?;
+
+        if let TxOpResponse::Single(resp) = msg {
+            resp
+        } else {
+            Err(self.handle_error(msg).into())
+        }
+    }
+
+    async fn batch_execute(
+        &self,
+        operations: Vec<Operation>,
+        trace_id: Option<String>,
+    ) -> crate::Result<Vec<crate::Result<ResponseData>>> {
+        let msg_req = TxOpRequestMsg::Batch(operations, trace_id);
+
+        let msg = self.send_and_receive(msg_req).await?;
+
+        if let TxOpResponse::Batch(resp) = msg {
+            resp
+        } else {
+            Err(self.handle_error(msg).into())
+        }
+    }
+
+    async fn send_and_receive(&self, msg: TxOpRequestMsg) -> Result<TxOpResponse, crate::CoreError> {
+        let (rx, req) = self.create_rx_and_req(msg);
```

Comment:
`Rx` naming wasn't super clear to me and I needed to scroll a bit to find out what it is. I can imagine it's also not clear to others when reading this.